File: LeetCode Solutions/LeetCode 75/Level 2/Day 7 Tree/Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0
        def height(root):
            if not root: 
                return 0
            left, right = height(root.left), height(root.right)
            self.diameter = max(self.diameter, left + right)
            return 1 + max(left, right)
        height(root)
        return self.diameter

File: LeetCode Solutions/October 2021 LeetCoding Challenge/Diameter of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        maxheight(root) ;
        return diameter;
    }
    private int maxheight(TreeNode root) {
        if (root == null){
            return 0;
        }
        int left = maxheight(root.left);
        int right = maxheight(root.right);    
        diameter = Math.max(diameter , left + right);
        return Math.max(left, right) + 1;
    }
}